### PR TITLE
Initial (basic) WebSockets support

### PR DESCRIPTION
**Pull request by fwsGonzo**
Unfortunately, lacks ability to close websocket when about to go idle. Needs help from TCP from a future reworking of how on_write works.

### CODE REVIEW

**Comment by fwsGonzo** [comment]
jenkins test please

File: api/net/websockets/websocket.hpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+#ifndef NET_WEBSOCKETS_WEBSOCKET_HPP
+#define NET_WEBSOCKETS_WEBSOCKET_HPP
+
+#include <net/http/server.hpp>
```
Comment (AndreasAakesson) [inline api/net/websockets/websocket.hpp]: I don't see `Server` referenced, maybe `<net/http/response_writer.hpp>` is enough.

File: src/net/websockets/websocket.cpp [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+}
+
+WebSocket::WebSocket(tcp::Connection_ptr tcpconn, bool client)
+  : conn(tcpconn), clientside(client)
+{
+  assert(conn != nullptr);
+}
+WebSocket::WebSocket(WebSocket&& other)
+{
+  printf("MOVED\n");
```
Comment (AndreasAakesson) [inline src/net/websockets/websocket.cpp]: printf ;)

File: src/net/websockets/websocket.cpp [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+  http::Header_set ws_headers {
+      {"Origin",     origin  },
+      {"Host",       remote.str()},
+      {"Connection", "Upgrade"  },
+      {"Upgrade",    "WebSocket"},
+      {"Sec-WebSocket-Version", "13"},
+      {"Sec-WebSocket-Key",     key }
+  };
+  // send HTTP request
+  printf("Sending request with key %s (len=%u)\n", key.c_str(), key.size());
```
Comment (AndreasAakesson) [inline src/net/websockets/websocket.cpp]: printf, there is also one on failure.

**Pull request closed without merging**